### PR TITLE
Our readme should be clearer about our products.

### PR DESCRIPTION
**Pull request by Lukasa**
Motivation:

We should explain when we want users to use specific products, so that
they can understand what to import and when.

Modifications:

- Added some text to the README to clarify what modules exist.

Result:

Should be easier to understand what there is in the NIO ecosystem

### CODE REVIEW

**Review by glbrntt — APPROVED** [review]
LGTM

File: README.md [hunk 1]
```diff
@@ -22,6 +22,18 @@ Repository | NIO 2 (Swift 5.2+) | NIO 1 (Swift 4+)
 
 NIO 2.29.0 and older support Swift 5.0+.
 
+Within this repository we have a number of products that provide different functionality. This package contains the following products:
```
[the remark this answers is not in the document]
Comment (Lukasa) [reply README.md:25]: I don't think I agree. My inclination is to say that the table above indicates what each of the various locations contains: then we go into details about what specifically is in this package.

File: README.md [hunk 2]
```diff
@@ -92,14 +104,14 @@ The following sections will describe the low-level tools that SwiftNIO provides,
...
-- [`ChannelHandler`][ch], a protocol
-- `Bootstrap`, several related structures
-- [`ByteBuffer`][bb], a struct
-- [`EventLoopFuture`][elf], a generic class
-- [`EventLoopPromise`][elp], a generic struct.
+- [`EventLoopGroup`][elg], a protocol, provided by `NIOCore`.
+- [`EventLoop`][el], a protocol, provided by `NIOCore`.
+- [`Channel`][c], a protocol, provided by `NIOCore`.
+- [`ChannelHandler`][ch], a protocol, provided by `NIOCore`.
+- `Bootstrap`, several related structures, provided by `NIO`.
```
[the remark this answers is not in the document]
Comment (Lukasa) [reply README.md:111]: Yes.

File: README.md [hunk 3]
```diff
@@ -109,7 +121,7 @@ The basic I/O primitive of SwiftNIO is the event loop. The event loop is an obje
...
+In SwiftNIO today there is one [`EventLoopGroup`][elg] implementation, and two [`EventLoop`][el] implementations. For production applications there is the [`MultiThreadedEventLoopGroup`][mtelg], an [`EventLoopGroup`][elg] that creates a number of threads (using the POSIX [`pthreads`][pthreads] library) and places one `SelectableEventLoop` on each one. The `SelectableEventLoop` is an event loop that uses a selector (either [`kqueue`][kqueue] or [`epoll`][epoll] depending on the target system) to manage I/O events from file descriptors and to dispatch work. These [`EventLoop`s][el] and [`EventLoopGroup`s][elg] are provided by the `NIO` module. Additionally, there is the [`EmbeddedEventLoop`][eel], which is a dummy event loop that is used primarily for testing purposes, provided by the `NIOEmbedded` module.
```
[the remark this answers is not in the document]
Comment (Lukasa) [reply README.md:124]: I'm somewhat inclined not to: given that we have the link out above, I don't think any additional links out from here add much value, and I worry about overwhelming folks. This section already does an awful lot.

**Pull request merged**